src/result-async.ts

```diff
@@ -0,0 +1,123 @@
+import { Result, Ok, Err } from './'
+import { isPromise } from './'
+
+export class ResultAsync<T, E> {
+  private _promise: Promise<Result<T, E>>
+
+  constructor(res: Promise<Result<T, E>>) {
+    this._promise = res
+  }
+
+  static fromPromise<T, E>(promise: Promise<T>): ResultAsync<T, E> {
+    return new ResultAsync(promise.then((value: T) => new Ok<T, E>(value)))
+  }
+
+  map<A>(_f: (t: T) => A | Promise<A>): ResultAsync<A, E> {
+    const newPromise: Promise<Result<A, E>> = this._promise.then((res: Result<T, E>) => {
+      if (res.isErr()) {
+        return new Err<A, E>(res.error)
+      }
+
+      const newValue = _f(res.value)
+
+      if (isPromise(newValue)) {
+        return newValue.then(value => new Ok<A, E>(value))
+      }
+
+      return new Ok<A, E>(newValue)
+    })
```

Comment:
Looks great! I have a few thoughts. 

For the sake of keeping code as legible as possible:
- Any reason not to use `async` / `await` syntax? It might lead to less indentation
- Please omit explicit type annotations if the compiler can infer the type already

The code on line 16 would look like this:  

```typescript
    const newPromise = this._promise.then((result) => { ...
```

Secondly, the instantiation of `Err` and `Ok` values on line 18, 24 and 27 shouldn't need to be annotated with their respective type arguments either. See if you can omit them. The compiler should be able to infer the types in all of these cases. 

If you do go ahead with these changes, please apply them to all the methods below as well! Otherwise please respond with some counterarguments so we can agree on a consistent coding style!

Reply:
> * Any reason not to use `async` / `await` syntax? It might lead to less indentation

These methods need to be remain synchronous (as they are returning a `ResultAsync` instance) and `async` / `await` is only available inside asynchronous functions.

However, it's possible to pass an asynchronous function to `.then()` and `await` the value instead of chaining with an addition `.then()`. 
I also noticed, we can `await` non-promise values as well. Given we don't know if the provided function returns a promise or not, we can `await` in both cases.

Looks a lot cleaner like this:
```typescript
  map<A>(f: (t: T) => A | Promise<A>): ResultAsync<A, E> {
    const newPromise = this._promise.then(async (res: Result<T, E>) => {
      if (res.isErr()) {
        return new Err<A, E>(res.error)
      }

      return new Ok<A, E>(await f(res.value))
    })

    return new ResultAsync(newPromise)
  }
```

Added bonus: we don't need `isPromise` at all anymore 🎉